Updated authentication doc JWT example

AndreYonadam opened the pull request:
Updated the authentication doc JWT example to include an example of generating and validating a token.

siemensikkema commented:
@AndreYonadam thanks! I think going for a more realistic example is an improvement, for sure. Renaming from `TestUser` to `SessionToken` is also better IMO because you wouldn't normally have a full user model inside a JWT.

A few thoughts:
- using `req.auth.require()` in the login endpoint assumes the user is already logged in so it would be better to show you can extract credentials from the request, find a user from the db, and verify the credentials (maybe just as a comment to keep the example clean) and then return the token.
- returning a dictionary for the token instead of an Encodable struct might set the "wrong" example although it might be fine because it shortens the example

AndreYonadam commented:
@siemensikkema Thanks for the feedback! I agree with both of your statements. I plan on amending this PR tomorrow based on your feedback.

AndreYonadam commented:
@siemensikkema I hope my updates addressed your concerns. I created a simple model to return instead of the dictionary and also offered two examples for login endpoints, one that tries to show a more "manual" approach of creating a SessionToken.

siemensikkema reviewed (commented):
Since I last looked at your PR I have had some experience with a situation where a user is more or less fully represented in a token. So having a `TestUser` that is also a `JWTPayload` as the original documentation suggests is realistic scenario (to contradict what I commented last time).

In my case this was using an external authentication provider (Keycloak) but you could build it like this all in your Vapor app. The question here is to choose what scenario to use for a concise but clear and somewhat realistic example.

I'm not sure the "validateLoggedInUser" route adds much on top of the current "me" route example which also validates the token. When it comes to an example of _generating_ a token maybe all we need to do is to link to [this section](https://docs.vapor.codes/4.0/jwt/#signing) along the lines of:

> See [here](https://docs.vapor.codes/4.0/jwt/#signing) for documentation on how to generate a token.

AndreYonadam commented on 4.0/docs/authentication.md:
````diff
@@ -748,35 +748,80 @@ The `CredentialsAuthenticator` extracts the `username` and `password` from the r
+Using our model for the JWT token and response, we can use a password protected login route which returns a `ClientTokenReponse` and includes a signed `SessionToken`.
+
+```swift
+let passwordProtected = app.grouped(User.authenticator())
````
> Are you proposing to handle verifying login credentials using an authenticator?
That was my intention here. If you refer to the example above in the "Model Authenticatable" section, I believe it verifies login credentials the same way, which I feel like is more structured then handling it manually in the route handler itself. Isn't it better practice to use `ModelAuthenticatable`?

With that being said, an advantage in handling verifying the login credentials in the route handler itself for the purposes of this example is in case the user doesn't know about `ModelAuthenticatable` yet.

0xTim replied:
Authenticators don't necessarily protect routes. They take some data (login credentials, auth token, JWT etc) and convert that into an authenticated type in the authentication cache. In this instance the route is only password protected because of the call to `try req.auth.require(User.self)`. We should probably add the guard middleware to make this clear

```suggestion
let passwordProtected = app.grouped(User.authenticator(), User.guardMiddleware())
```

AndreYonadam commented on 4.0/docs/authentication.md:
````diff
@@ -748,35 +748,80 @@ The `CredentialsAuthenticator` extracts the `username` and `password` from the r
+}
+```
+
+Alternatively, if you don't want to use a an authenticator you can have something that looks like the following.
````
Interesting. I recently read this authentication document not to long ago when I was new to Vapor. I feel like the "Model Authenticator" section gives off the impression that an authenticator is used for validating user credentials. That section has an example about confirming a User model to `ModelAuthenticatable` and uses it to password-protect the login route. In fact, an example request was provided that contains an Authorization header.

I'm pretty new to Vapor and I feel like if that's not what an authenticator is for then other new users might have similar confusion.

0xTim commented on 4.0/docs/authentication.md:
```diff
@@ -748,35 +748,80 @@ The `CredentialsAuthenticator` extracts the `username` and `password` from the r
 [JWT](jwt.md) provides a `JWTAuthenticator` that can be used to authenticate JSON Web Tokens in incoming requests. If you are new to JWT, check out the [overview](jwt.md).
 
-First, create a type representing the structure of the JWT payload.
+First, create a struct representing a JWT payload.
```
Nit - type is the technically correct term here

```suggestion
First, create a type representing a JWT payload.
```

0xTim reviewed (changes requested):
Couple of minor changes then good to go in my opinion - @siemensikkema any else from your end?

AndreYonadam commented:
> Couple of minor changes then good to go in my opinion - @siemensikkema any else from your end?

@0xTim Thanks for your suggestions. Feel free to let me know if there is anything else I can change.

0xTim reviewed (approved):
Thanks!
